src/Interpreters/Cache/QueryCache.h

```diff
@@ -51,8 +51,16 @@ class QueryCache
         /// Result metadata for constructing the pipe.
         const Block header;
 
-        /// The user who executed the query.
+        /// The name, id and current roles of the user who executed the query.
+        /// These members are necessary to ensure that a (non-shared, see below) entry can only be written and read by the same user with
+        /// the same roles. Example attack scenarios:
+        /// - after DROP USER, it must not be possible to create a new user with with the dropped user name and access the dropped user's
+        ///   query cache entries
+        /// - different roles of the same user may be tied to different row-level policies. It must not be possible to switch role and
+        ///   access another role's cache entries
         const String user_name;
```

Comment:
Is `user_name` still used?

Reply:
You are right. But before fixing this: do you know why the `user_id` / `getUserID()` in `Context` is an `std::optional`? When would the user id be not set?